README.md
# MXETHControl
ELV/EQ3 ETH components to MQTT Bridge

Runs on a Wemos D1 mini/ESP8266, and uses an RFM69HCW module to capture
packages from ELV ETH 200 comfort "window sensors" and "remote controls"
and publishes them via MQTT.
It also subscribes to MQTT topics to control thermostat devices by
simulating one "remote control" per thermostat.

It also includes firmware OTA updates either checked at every boot or
triggered via MQTT.

For more details see the readme.txt file.


changelog.txt
### Changelog
2020-11-29  0.1  First version, just with WiFi, MQTT etc. nothing RFM69 yet
2020-12-05  0.2-0.4 A lot of progress
2020-12-13  0.5  Datapakets are now properly received via RFM69 decoded and verified (CRC)
2020-12-14  0.6  sensor raw packages are now analyzed and being sent to MQTT
2020-12-14  0.7  Added publishing of the complete sensor package as json string
2020-12-25  0.8  Added first subscription to thermostat topics
2020-12-25  0.9  Added packet assembling, crc calculation, reverse byte order, bit stuffing
2020-12-26  1.0  Added packet sending, it's not yet properly received by the thermostat
2021-01-10  1.3  Contains partially usage of a custom preamble. Not working.
2021-01-10  1.4  Changed TX to send unlimited bit stream with own preamble and own Sync words
2021-01-10  1.5  WindowSensor and RemoteControl work almost fine. Only issue are negative numbers
                 which means where bit stuffing has happened. So in general bit stuffed commands are broken.
2021-01-10  1.6  Fully working version. TODO: Add queuing mechanism for MQTT cmds.
2021-01-11  1.7  Added float verification of temperature setting
                 Added MQTT command queue.
                 Removed useless USE_MQTT define
2021-01-11  1.8  Implemented RFM69 reset via MQTT
                 Code cleanup
2021-01-11  1.9  Changed CRC error message to MXDEBUG level.
2021-01-14  2.0  Added CFG_RF69_POWERLEVEL; Added two more thermostats
2021-01-17  2.1  Increased CFG_RF69_POWERLEVEL from 25 to 31, sometime ThermostatSZ does not get the messages.
2021-01-30  2.3  Added 99xxxx group topics, they are exactly handled like thermostats,
                 it's just to put those at the end of the thermostat list.
2021-02-06  2.4  Restructured mqtt topics and removed several retain flags.
                 Implemented status/state topic.
                 Implemented bugfix to ignore empty thermostat commands.
2021-02-06  2.5  Fixed bug that "online" state is not properly retained
2021-02-13  2.6  Code cleanup



src/ETH200RFM69.h
/****************************************************************************
ETH200RFM69 modification for ELV ETH200

For use with an RFM69CW module and specific behavior for the ELV ETH200
sensors and actors.  

Copyright 2020 mt-mrx <64284703+mt-mrx@users.noreply.github.com>

Based on:
  RFM69 library:
    https://github.com/LowPowerLab/RFM69
*****************************************************************************
License
*****************************************************************************
This program is free software; you can redistribute it 
and/or modify it under the terms of the GNU General    
Public License as published by the Free Software       
Foundation; either version 3 of the License, or        
(at your option) any later version.                    
                                                        
This program is distributed in the hope that it will   
be useful, but WITHOUT ANY WARRANTY; without even the  
implied warranty of MERCHANTABILITY or FITNESS FOR A   
PARTICULAR PURPOSE. See the GNU General Public        
License for more details.                              
                                                       
Licence can be viewed at                               
http://www.gnu.org/licenses/gpl-3.0.txt

Please maintain this license information along with authorship
and copyright notices in any redistribution of this code
****************************************************************************/

#ifndef ETH200RFM69_h
  #define ETH200RFM69_h

  #include <RFM69.h>
  #include <RFM69registers.h>

  class ETH200RFM69: public RFM69 {
    public:
      static uint8_t PAYLOADETH200;
      static uint16_t ETH200CRCStartWindowSensor;
      static uint16_t ETH200CRCStartRemoteControl;
      static uint16_t ETH200CRCMask;
      uint8_t currentPacketCounter = 1; // the current packet counter, will be incremented whenever a packet is sent
      uint8_t *lastSentPacket; // the last raw packet we sent out, pointer to an array which will be initialized during constructor
      uint8_t lastSentPacketSize = 0;
      ETH200RFM69(uint8_t slaveSelectPin=RF69_SPI_CS, uint8_t interruptPin=RF69_IRQ_PIN, bool isRFM69HW=false); //override
      bool initialize(); //override
      void readAllRegs(); //override
      virtual bool receiveDone(); //override
      boolean send(uint8_t buffer[], uint8_t bufferSize, uint8_t numStuffedBits = 0); //override and signature change
      boolean sendPacket(uint8_t deviceType, uint32_t address, uint8_t cmd, uint8_t cmds[], uint8_t cmdsSize); // sends a packet
    protected:
      static void isr0(); //override
      void interruptHandler(); //override
      void sendFrame(uint8_t buffer[], uint8_t bufferSize, uint8_t numStuffedBits = 0); //override and signature change
      uint8_t reverseByte(uint8_t b);
      uint16_t calcCRC16r(uint16_t c,uint16_t crc, uint16_t mask);
      uint16_t calcPacketCRC16r(uint8_t packet[], uint8_t length, uint16_t crcStart, uint16_t crcMask);
      uint8_t destuffPayload(uint8_t inBuf[], uint8_t destuffedBuf[], uint8_t length);
      uint8_t stuffPayload(uint8_t inBuf[], uint8_t stuffedBuf[], uint8_t length);
  };
#endif // ETH200RFM69_h

src/MXDebugUtils.h
/****************************************************************************
MXDebugUtils.h - Simple debugging utilities.
 
Copyright 2020 mt-mrx <64284703+mt-mrx@users.noreply.github.com>

Based on ideas taken from:
  http://www.arduino.cc/cgi-bin/yabb2/YaBB.pl?num=1271517197

*****************************************************************************
License
*****************************************************************************
This program is free software; you can redistribute it 
and/or modify it under the terms of the GNU General    
Public License as published by the Free Software       
Foundation; either version 3 of the License, or        
(at your option) any later version.                    
                                                        
This program is distributed in the hope that it will   
be useful, but WITHOUT ANY WARRANTY; without even the  
implied warranty of MERCHANTABILITY or FITNESS FOR A   
PARTICULAR PURPOSE. See the GNU General Public        
License for more details.                              
                                                       
Licence can be viewed at                               
http://www.gnu.org/licenses/gpl-3.0.txt

Please maintain this license information along with authorship
and copyright notices in any redistribution of this code
****************************************************************************/

#ifndef MXDEBUGUTILS_H
  #define MXDEBUGUTILS_H
  #include <config.h>            // project settings file, need to include to use MXDEBUG

  // Debug print functions
  // and why they need to be static:
  // https://stackoverflow.com/questions/34997795/inline-function-multiple-definition
  
  static void mxDebugPrint(const char* prefix, const char* func, const char* file,
                           int line, const char* text, bool printloc) __attribute__((unused));
  static void mxDebugPrint(const char* prefix, const char* func, const char* file,
                           int line, const char* text, bool printloc) {
    if (printloc) {
      //print location information of the calling statement
      Serial.print(prefix);
      Serial.print(millis());
      Serial.print(": ");
      Serial.print(func);
      Serial.print(" ");
      Serial.print(file);
      Serial.print(":");
      Serial.print(line);
      Serial.print(" ");
    } 
    Serial.print(text);
  }

  static void mxDebugPrint(const char* prefix, const char* func, const char* file,
                           int line, int text, bool printloc) __attribute__((unused));
  static void mxDebugPrint(const char* prefix, const char* func, const char* file,
                           int line, int text, bool printloc) {
    mxDebugPrint(prefix, func, file, line, String(text).c_str(), printloc);
  }

  static void mxDebugPrint(const char* prefix, const char* func, const char* file,
                           int line, String text, bool printloc) __attribute__((unused));
  static void mxDebugPrint(const char* prefix, const char* func, const char* file,
                           int line, String text, bool printloc) {
    mxDebugPrint(prefix, func, file, line, text.c_str(), printloc);
  }

  #ifdef MXDEBUG
    #define MXDEBUG_PRINT(text) \
      mxDebugPrint("DEBUG: ", __PRETTY_FUNCTION__, __FILE__, __LINE__, text, false);
      //print text if MXDEBUG enabled without newline
    #define MXDEBUG_PRINTLN(text) \
      mxDebugPrint("DEBUG: ", __PRETTY_FUNCTION__, __FILE__, __LINE__, text, false); \
      Serial.println();
      //print text if MXDEBUG enabled with newline
    #define MXDEBUG_PRINTL(text) \
      mxDebugPrint("DEBUG: ", __PRETTY_FUNCTION__, __FILE__, __LINE__, text, true);
      //print text if MXDEBUG enabled without newline, with call location
    #define MXDEBUG_PRINTLLN(text) \
      mxDebugPrint("DEBUG: ", __PRETTY_FUNCTION__, __FILE__, __LINE__, text, true); \
      Serial.println();
      //print text if MXDEBUG enabled with newline, with call location
  #else
    // The non-debug macro means that the code is validated but never called.
    // See chapter 8 of 'The Practice of Programming', by Kernighan and Pike.
    // source: https://stackoverflow.com/questions/1644868/define-macro-for-debug-printing-in-c
    #define MXDEBUG_PRINT(text) \
      do { if (0) mxDebugPrint("", "", "", 0, text, true); } while (0);
    #define MXDEBUG_PRINTLN(text) \
      do { if (0) mxDebugPrint("", "", "", 0, text, true); } while (0);
    #define MXDEBUG_PRINTL(text) \
      do { if (0) mxDebugPrint("", "", "", 0, text, true); } while (0);
    #define MXDEBUG_PRINTLLN(text) \
      do { if (0) mxDebugPrint("", "", "", 0, text, true); } while (0);
  #endif //MXDEBUG

  #ifdef MXINFO
    #define MXINFO_PRINT(text) \
      mxDebugPrint("INFO: ", __PRETTY_FUNCTION__, __FILE__, __LINE__, text, false);
      //print text if MXINFO enabled without newline
    #define MXINFO_PRINTLN(text) \
      mxDebugPrint("INFO: ", __PRETTY_FUNCTION__, __FILE__, __LINE__, text, false); \
      Serial.println();
      //print text if MXINFO enabled with newline
    #define MXINFO_PRINTL(text) \
      mxDebugPrint("INFO: ", __PRETTY_FUNCTION__, __FILE__, __LINE__, text, true);
      //print text if MXINFO enabled without newline, with call location
    #define MXINFO_PRINTLLN(text) \
      mxDebugPrint("INFO: ", __PRETTY_FUNCTION__, __FILE__, __LINE__, text, true); \
      Serial.println();
      //print text if MXINFO enabled with newline, with call location
  #else
    // The non-info macro means that the code is validated but never called.
    // See chapter 8 of 'The Practice of Programming', by Kernighan and Pike.
    // source: https://stackoverflow.com/questions/1644868/define-macro-for-debug-printing-in-c
    #define MXINFO_PRINT(text) \
      do { if (0) mxDebugPrint("", "", "", 0, text, true); } while (0);
    #define MXINFO_PRINTLN(text) \
      do { if (0) mxDebugPrint("", "", "", 0, text, true); } while (0);
    #define MXINFO_PRINTL(text) \
      do { if (0) mxDebugPrint("", "", "", 0, text, true); } while (0);
    #define MXINFO_PRINTLLN(text) \
      do { if (0) mxDebugPrint("", "", "", 0, text, true); } while (0);
  #endif //MXINFO

  // timinig code execution parts
  // and why they need to be static:
  // https://stackoverflow.com/questions/34997795/inline-function-multiple-definition
  static unsigned long tsLast;  //millis() value for last MXTIME_PRINT() function call
  static void mxDebugTime(const char* prefix, const char* func, const char* file,
                          int line, const char* text) __attribute__((unused));
  static void mxDebugTime(const char* prefix, const char* func, const char* file,
                          int line, const char* text) {
    unsigned long tsCurrent = millis();
    unsigned long tsDelta = tsCurrent - tsLast;
    Serial.print(prefix);
    Serial.print(func);
    Serial.print(' ');
    Serial.print(file);
    Serial.print(':');
    Serial.print(line);
    Serial.print(F(" Time Since Start: "));
    Serial.print(tsCurrent);
    Serial.print(F("ms, Time Since Last MXTIME_PRINT(): "));
    Serial.print(tsDelta);
    Serial.println(F("ms "));
    Serial.println(text);
    //setting last call of this function
    tsLast = millis();
  }

  static void mxDebugTime(const char* prefix, const char* func, const char* file,
                          int line, int text) __attribute__((unused));
  static void mxDebugTime(const char* prefix, const char* func, const char* file,
                          int line, int text) {
    mxDebugTime(prefix, func, file, line, String(text).c_str());
  }

  static void mxDebugTime(const char* prefix, const char* func, const char* file,
                          int line, String text) __attribute__((unused));
  static void mxDebugTime(const char* prefix, const char* func, const char* file,
                          int line, String text) {
    mxDebugTime(prefix, func, file, line, text.c_str());
  }

  #ifdef MXDEBUG_TIME
    #define MXTIME_PRINT(text) \
      mxDebugTime("TIME: ", __PRETTY_FUNCTION__, __FILE__, __LINE__, text);
  #else
    // The non-time macro means that the code is validated but never called.
    // See chapter 8 of 'The Practice of Programming', by Kernighan and Pike.
    // source: https://stackoverflow.com/questions/1644868/define-macro-for-debug-printing-in-c
    #define MXTIME_PRINT(text) \
      do { if (0) mxDebugTime("", "", "", 0, text); } while (0)
  #endif //MXDEBUG_TIME

  /*
    Printing binary and hex values with leading zeros
    Author: septillion
    Source: https://forum.arduino.cc/index.php?topic=464778.0
    and why they need to be static:
    https://stackoverflow.com/questions/34997795/inline-function-multiple-definition
  */
  static void printBinWithZeroPad(Print &out, unsigned long n, byte size) {
    for (byte i = size - 1; ;i--) {
      out.print(bitRead(n, i));

      // need to break at exactly 0 and cannot use condition like i>=0 because
      // if you reach i=0 and substract 1 the result is then 255 and the
      // condition i>=0 would be true again
      if(i == 0) {
        break;
      }
    }
  }
  static inline void printBinWithZeroPad(Print &out, unsigned long n) { printBinWithZeroPad(out, n, sizeof(n) * 8); }
  static inline void printBinWithZeroPad(Print &out, unsigned int n) { printBinWithZeroPad(out, n, sizeof(n) * 8); }
  static inline void printBinWithZeroPad(Print &out, uint16_t n) { printBinWithZeroPad(out, n, sizeof(n) * 8); }
  static inline void printBinWithZeroPad(Print &out, uint8_t n) { printBinWithZeroPad(out, n, sizeof(n) * 8); }

  static void printHexWithZeroPad(Print &out, unsigned long n, byte size) {
    for(byte i = size - 1; ;i--) {
      out.print((n >> (i * 4)) & 0x0F, HEX);
    
      // see printBinWithZeroPad for explanation
      if(i == 0) {
        break;
      }
    }
  }
  static inline void printHexWithZeroPad(Print &out, unsigned long n) { printHexWithZeroPad(out, n, sizeof(n) * 2); }
  static inline void printHexWithZeroPad(Print &out, unsigned int n) { printHexWithZeroPad(out, n, sizeof(n) * 2); }
  static inline void printHexWithZeroPad(Print &out, uint16_t n) { printHexWithZeroPad(out, n, sizeof(n) * 2); }
  static inline void printHexWithZeroPad(Print &out, byte n) { printHexWithZeroPad(out, n, sizeof(n) * 2); }
#endif //MXDEBUGUTILS_H


src/MXPubSubClientWrapper.h
/****************************************************************************
MXPubSubClientWrapper.h - Simple extension of PubSubClient.

Includes functions which can handle the String class.
 
Copyright 2020 mt-mrx <64284703+mt-mrx@users.noreply.github.com>

Based on:
 PubSubClient - A simple client for MQTT.
  Nick O'Leary
  http://knolleary.net
*****************************************************************************
License
*****************************************************************************
This program is free software; you can redistribute it 
and/or modify it under the terms of the GNU General    
Public License as published by the Free Software       
Foundation; either version 3 of the License, or        
(at your option) any later version.                    
                                                        
This program is distributed in the hope that it will   
be useful, but WITHOUT ANY WARRANTY; without even the  
implied warranty of MERCHANTABILITY or FITNESS FOR A   
PARTICULAR PURPOSE. See the GNU General Public        
License for more details.                              
                                                       
Licence can be viewed at                               
http://www.gnu.org/licenses/gpl-3.0.txt

Please maintain this license information along with authorship
and copyright notices in any redistribution of this code
****************************************************************************/

#ifndef MXPUBSUBCLIENTWRAPPER_H
    #define MXPUBSUBCLIENTWRAPPER_H
  /*
  * #define MQTT_MAX_PACKET_SIZE 256
  * increases MQTT message sizes, needs to be done either in the PubSubClient
  * library or can be done in platformio.ini file.
  * If you would define it here, it has no effect on the library
  * I think since version 2.8 there also exist a method setBufferSize() to
  * override MQTT_MAX_PACKET_SIZE
  */
  #include <PubSubClient.h>      // for MQTT

  class MXPubSubClientWrapper : public PubSubClient {
    private:
    public:
      MXPubSubClientWrapper(Client& espc);
      bool publish(StringSumHelper topic, String str);
      bool publish(StringSumHelper topic, unsigned int num);
      bool publish(const char* topic, String str);
      bool publish(const char* topic, unsigned int num);

      bool publish(StringSumHelper topic, String str, bool retain);
      bool publish(StringSumHelper topic, unsigned int num, bool retain);
      bool publish(const char* topic, String str, bool retain);
      bool publish(const char* topic, unsigned int num, bool retain);
  };

  MXPubSubClientWrapper::MXPubSubClientWrapper(Client& espc) : PubSubClient(espc) {

  }

  bool MXPubSubClientWrapper::publish(StringSumHelper topic, String str) {
    return publish(topic.c_str(), str);
  }

  bool MXPubSubClientWrapper::publish(StringSumHelper topic, unsigned int num) {
    return publish(topic.c_str(), num);
  }

  bool MXPubSubClientWrapper::publish(const char* topic, String str) {
    return publish(topic, str, false);
  }

  bool MXPubSubClientWrapper::publish(const char* topic, unsigned int num) {
    return publish(topic, num, false);
  }

  bool MXPubSubClientWrapper::publish(StringSumHelper topic, String str, bool retain) {
    return publish(topic.c_str(), str, retain);
  }

  bool MXPubSubClientWrapper::publish(StringSumHelper topic, unsigned int num, bool retain) {
    return publish(topic.c_str(), num, retain);
  }

  bool MXPubSubClientWrapper::publish(const char* topic, String str, bool retain) {
    char buf[128];

    if (str.length() >= 128) return false;

    str.toCharArray(buf, 128);
    return PubSubClient::publish(topic, buf, retain);
  }

  bool MXPubSubClientWrapper::publish(const char* topic, unsigned int num, bool retain) {
    char buf[6];

    dtostrf(num, 0, 0, buf);
    return PubSubClient::publish(topic, buf, retain);
  }
#endif //MXPUBSUBCLIENTWRAPPER_H

src/config.h
/*
 * config.h - General config file for MXETHControl
 */

#ifndef MXETHCONTROL_CONFIG_H
    #define MXETHCONTROL_CONFIG_H

    /*** Begin: WiFi Settings ***/
    #define CFG_WIFI_SSID "<changeme>"
    #define CFG_WIFI_PASSWORD "<changeme>"
    /*** End: WiFi Settings ***/

    // Set serial output verbosity
    //#define MXDEBUG                  // to enable debugging output, does not include MXINFO
    #define MXINFO                   // to enable info output
    //#define MXDEBUG_TIME           // to enable timing debugging output

    /*** Begin: Firmware Update settings ***/
    // if defined check/apply for firmware update via HTTP at every wake up/reboot, adds ~400ms
    //#define HTTP_OTA_FW_UPD
    // if defined do an HTTP firmware check/update, when triggered by an
    // MQTT MXETHControl/<MAC>/set/update "1" message
    #define MQTT_HTTP_OTA_FW_UPD
    // The MX ETH Control device checks that folder for two files based on its MAC address
    // firmware binary  - <deviceName>.bin     , e.g. MXETHControl807D3A78FF16.bin
    // firmware version - <deviceName>.version , e.g. MXETHControl807D3A78FF16.version
    #define CFG_FW_BASE_URL "http://192.168.199.10/MXETHControl/firmware/"
    // Firmware version, should match the changelog.txt and is used for OTA firmware
    // updates
    #define CFG_FW_VERSION "2.6"
    /*** End: Firmware Update settings ***/

    /*** Begin: MQTT settings ***/
    #define CFG_MQTT_SERVER {192, 168, 199, 79}  // andromeda
    #define CFG_MQTT_PORT 1883                   // default non ssl port is 1883
    #define CFG_MQTT_USER "<changeme>"
    #define CFG_MQTT_PASSWORD "<changeme>"
    /*** End: MQTT settings ***/

    /*** Begin: PIN settings ***/
    /*
    ### Pinout
    # ESP8266       WemosD1Mini    RFM69(H)CW   CW-Pin#   HCW-Pin#
    # GPIO15/SS/CS  D8             NSS/CS       7         5
    # GPIO14/SCLK   D5             SCK/CLK      6         4
    # GPIO4         D2             DIO0         9         14        # Interrupt
    # GPIO12/MISO   D6             MISO         8         2
    # GND           G              GND          3         1
    # GND           G              GND          14        8
    # GND           G              GND          N/A       10       
    # GPIO13/MOSI   D7             MOSI         5         3
    # GPIO0         D3             Reset        13        6         # Hard Reset of RFM module
    #               3V3            VCC          2         13
    #                              Ant          1         9         # Antenna
    */
    #define CFG_RF69_IRQ_PIN          4   // I first tried GPIO2 but that caused the Wemos to reset
    #define CFG_RF69_SPI_CS           15
    #define CFG_RF69_ISRFM69HW        true   // is required for RFM69H(C)W otherwise TX does not work
    #define CFG_RF69_POWERLEVEL       31     //min:0 - max:31, transmission power
    /* A manual reset of the RFM69CW is possible even for applications in which
       VDD cannot be physically disconnected. The RESET pin should be pulled high
       for a hundred microseconds, and then released. The user should then wait
       for 5 ms before using the module.
    */
    #define CFG_RF69_RST_PIN          0
    /*** End: PIN settings ***/

    /*** Begin: misc settings ***/
    #define CFG_ETH200MAXPACKETSIZE 10 // should be max 10 bytes
    // number of times a packet is transmitted, should take ~6s, the thermostats don't
    // listen all the time it looks like they are checking only every ~5s. At least
    // the window sensor and the remote control send for ~6-8s their commands repeatedly.
    #define CFG_ETH200NUMPACKETSENDREPEATS 330 // 350,400 cause sometimes the thermostat to activate twice
    // max command bytes:
    //  remote control: 2
    //  windows sensor: 1
    #define CFG_ETH200MAXCMDS 2
    // number of thermostats to "create"
    #define CFG_ETH200NUMTHERMOSTATS 7  // 1 to 255
    // number of groups of thermostats to "create"
    #define CFG_ETH200NUMGROUPS 5       // 1 to 255

    // size of messages array to handle parallel incoming messages
    #define CFG_MESSAGES_SIZE 5
    // size of mqttCmds array to handle parallel incoming MQTT Cmds
    #define CFG_MQTTCMDS_SIZE 30
    // duration in seconds during which a message is received and how long we should wait before
    // sending it to MQTT. So if any external tool is reacting to that message and sending a new
    // command to the MXETHControl device we are sure we don't start sending if we still receive something.
    // "Remote Control" and "Windows Sensor" send their packages continuously for ~10 seconds
    #define CFG_MESSAGE_DELAY 13

    /*
      duration that the ESP uses as delay(x) at the end of each loop.
      It's a tradeoff between
        longer delay  - reduced power consumption
        shorter delay - faster response times (catches more packages)

      anyway, debug package output reduces the number of catched packages from ~170 -> ~26
      100ms, reduces power consumption from 88mA -> 40mA, with debug output enabled, reduces received packages 26 -> 14
      500ms, reduces power consumption from 88mA -> 32mA, with debug output enabled, reduces received packages 26 -> 5
    */
    #define CFG_ESP_LOOP_DELAY 100  // in ms
    //#define CFG_ESP_LOOP_DELAY 500

    // the duration that needs to expire after a (last) packet we got before updating
    // the MQTT status/state topic from "receiving" to "listening"
    // need to be a factor x longer than CFG_ESP_LOOP_DELAY
    #define CFG_STATE_RECEIVING_MAX_TIME 500  // in ms
    /*** End: misc settings ***/
#endif //MXETHCONTROL_CONFIG_H
